File: parse/src/main/java/com/parse/OfflineQueryController.java
/*
 * Copyright (c) 2015-present, Parse, LLC.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 */
package com.parse;

import java.util.List;

import bolts.Task;

class OfflineQueryController extends AbstractQueryController {

    private final OfflineStore offlineStore;
    private final ParseQueryController networkController;

    public OfflineQueryController(OfflineStore store, ParseQueryController network) {
        offlineStore = store;
        networkController = network;
    }

    @Override
    public <T extends ParseObject> Task<List<T>> findAsync(
            ParseQuery.State<T> state,
            ParseUser user,
            Task<Void> cancellationToken) {
        if (state.isFromLocalDatastore()) {
            return offlineStore.findFromPinAsync(state.pinName(), state, user);
        } else {
            return networkController.findAsync(state, user, cancellationToken);
        }
    }

    @Override
    public <T extends ParseObject> Task<Integer> countAsync(
            ParseQuery.State<T> state,
            ParseUser user,
            Task<Void> cancellationToken) {
        if (state.isFromLocalDatastore()) {
            return offlineStore.countFromPinAsync(state.pinName(), state, user);
        } else {
            return networkController.countAsync(state, user, cancellationToken);
        }
    }
}

File: parse/src/main/java/com/parse/OfflineQueryController.kt
/*
 * Copyright (c) 2015-present, Parse, LLC.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 */
package com.parse

import com.parse.boltsinternal.Task

internal class OfflineQueryController(
    private val offlineStore: OfflineStore,
    private val networkController: ParseQueryController
) : AbstractQueryController() {
    override fun <T : ParseObject> findAsync(
        state: ParseQuery.State<T>,
        user: ParseUser?,
        cancellationToken: Task<Void>?
    ): Task<List<T>> {
        return if (state.isFromLocalDatastore) {
            offlineStore.findFromPinAsync(state.pinName(), state, user)!!
        } else {
            networkController.findAsync(state, user, cancellationToken)
        }
    }

    override fun <T : ParseObject> countAsync(
        state: ParseQuery.State<T>,
        user: ParseUser?,
        cancellationToken: Task<Void>?
    ): Task<Int?> {
        return if (state.isFromLocalDatastore) {
            offlineStore.countFromPinAsync(state.pinName(), state, user)!!
        } else {
            networkController.countAsync(state, user, cancellationToken)
        }
    }
}